web_poet/requests.py

```diff
@@ -0,0 +1,142 @@
+
+You can read more about this in the :ref:`advanced-downloader-impl` documentation.
+"""
+
+import asyncio
+import logging
+from contextvars import ContextVar
+from typing import Optional, List, Dict, ByteString, Any, Union, Callable
+
+import attr
+
+from web_poet.page_inputs import HttpResponse
+
+logger = logging.getLogger(__name__)
+
+
+mapping = Dict[Union[str, ByteString], Any]
+
+# Frameworks that wants to support additional requests in ``web-poet`` should
+# set the appropriate implementation for requesting data.
+request_backend_var: ContextVar = ContextVar("request_backend")
+
+
+class RequestBackendError(Exception):
+    """Indicates that the ``web_poet.request_backend_var`` wasn't set
+    by the framework using **web-poet**.
+
+    See the documentation section about :ref:`setting up the contextvars <setup-contextvars>`
+    to learn more about this.
+    """
+
+    pass
+
+
+@attr.define
+class Request:
+    """Represents a generic HTTP request used by other functionalities in
+    **web-poet** like :class:`~.HttpClient`.
+    """
+
+    url: str
+    method: str = "GET"
+    headers: Optional[mapping] = None
+    body: Optional[str] = None
+
+
+async def _perform_request(request: Request) -> HttpResponse:
+    """Given a :class:`~.Request`, execute it using the **request implementation**
+    that was set in the ``web_poet.request_backend_var`` :mod:`contextvars`
+    instance.
+
+    .. warning::
+        By convention, this function should return a :class:`~.HttpResponse`.
+        However, the underlying downloader assigned in
+        ``web_poet.request_backend_var`` might change that, depending on
+        how the framework using **web-poet** implements it.
+    """
+
+    logger.info(f"Requesting page: {request}")
+
+    try:
+        request_backend = request_backend_var.get()
+    except LookupError:
+        raise RequestBackendError(
+            "Additional requests are used inside the Page Object but the "
+            "current framework has not set any Request Backend via "
+            "'web_poet.request_backend_var'"
+        )
+
+    response_data: HttpResponse = await request_backend(request)
+    return response_data
+
+
+class HttpClient:
+    """A convenient client to easily execute requests.
+
+    By default, it uses the request implementation assigned in the
+    ``web_poet.request_backend_var`` which is a :mod:`contextvars` instance to
+    download the actual requests. However, it can easily be overridable by
+    providing an optional ``request_downloader`` callable.
+
+    Providing the request implementation by dependency injection would be a good
+    alternative solution when you want to avoid setting up :mod:`contextvars`
+    like ``web_poet.request_backend_var``.
+
+    In any case, this doesn't contain any implementation about how to execute
+    any requests fed into it. When setting that up, make sure that the downloader
+    implementation returns a :class:`~.HttpResponse` instance.
```

Comment:
I just realized there is one big piece which is missing: we need to standardize the exception and error handling as well. Users should be able to write error handling code in POs, and it should work across backends. 

We should also document and test error handling.

```py
try:
    response = await self.http_client.get("...")
except ????:
    # error handling
```

So, the questions:

1. What are the exceptions raised?
2. When are they raised?

For example, there are connection errors, tls validation errors, read timeout errors, etc. There are also non-200 status codes. When are we raising an error, and when we're not?

Reply:
@kmike , as discussed offline, I've added the base exception that frameworks implementing **web-poet** should support. All other exceptions that we'll be adding and supporting in the future should subclass from it.

Let me know what you think! Here's the commit: https://github.com/scrapinghub/web-poet/pull/22/commits/eda3352139ea4ca62ad87835f73ff445185f33d3